### ISSUE TITLE
False-negative `unnecessary_cast` when making a type nullable.

### ISSUE THREAD

**Issue by naipofo** [dart-lang/linter#4792]
**Describe the issue**
Linter gives `unnecessary_cast` despite the cast fixing a type error.

**To Reproduce**

```dart
void main() {
  final list = [1,2,3,4];

  // type error: The return type 'Null' isn't a 'int', as required by the closure's context.
  final maybeFound1 = list.firstWhere((e) => e == 5, orElse: () => null);

  // error fixed but lint says that there is an unnecessary cast
  final maybeFound2 = (list as List<int?>).firstWhere((e) => e == 5, orElse: () => null);
}
```

**Expected behavior**
No lint message, as the cast is not unnecessary.

**Additional context**
Applying the lint fix as suggested causes a type error and dart refuses to compile.
